grizzly_extras/async_message/mq/rfh2.py

```diff
@@ -0,0 +1,157 @@
+
+    def _get_usr_encoding(self) -> Optional[str]:
+        if len(self.name_value_parts) == 0:
+            return None
+        for part in self.name_value_parts:
+            if part.tag == 'usr':
+                encoding_element = part.find('ContentEncoding')
+                if encoding_element is None:
+                    return None
+                return encoding_element.text
+        return None
+
+    def get_payload(self) -> bytes:
+        content_encoding = self._get_usr_encoding()
+        if content_encoding == 'gzip':
+            return gzip.decompress(self.payload)
+        return self.payload
+
+
+class Rfh2Encoder(object):
+    @classmethod
+    def create_md(cls) -> pymqi.MD:
+        md = pymqi.MD()
+        md.Format = 'MQHRF2  '.encode()
+        md.CodedCharSetId = 1208
+        md.Encoding = 546
+        return md
+
+    def __init__(self, payload: bytes, queue_name: str, encoding: str = 'gzip', tstamp: Optional[str] = None) -> None:
+        if encoding != 'gzip':
+            raise NotImplementedError('Only gzip encoding is implemented')
+        self.queue_name = queue_name
+        self.content_encoding = encoding
+        self.tstamp = tstamp
+        self.message = bytearray()
+
+        self._build_payload(payload)
+        self._build_name_values()
+        self._build_header()
+
+    def _build_payload(self, payload: bytes) -> None:
+        self.payload = gzip.compress(payload)
+
+    def _build_name_values(self) -> None:
+        padding_multiple = 4
+        self.name_values = bytearray()
+        content_length = len(self.payload)
+        if self.tstamp is None:
+            tstamp = str(round(time.time() * 1000))
+        else:
+            tstamp = self.tstamp
+        name_values_txt = [
+            "<mcd><Msd>jms_bytes</Msd></mcd>",
+            f"<jms><Dst>queue:///{self.queue_name}</Dst><Tms>{tstamp}</Tms><Dlv>2</Dlv></jms>",
+            f"<usr><ContentEncoding>{self.content_encoding}</ContentEncoding><ContentLength dt='i8'>{content_length}</ContentLength></usr>",
+        ]
+        for value in name_values_txt:
+            value_len = len(value)
+
+            # pad with spaces to get length to a multiple of 4
+            if value_len % padding_multiple != 0:
+                padding_length = padding_multiple - value_len % padding_multiple
+                value = value + ' ' * padding_length
+
+            name_value = value.encode()
+            name_value_length = len(name_value)
+            self.name_values.extend(pack('<l', name_value_length))
+            self.name_values.extend(name_value)
+
+    def _build_header(self) -> None:
+        header_length = 36
+        self.struc_id = 'RFH '
+        self.version = 2
+        self.struc_length = header_length + len(self.name_values)
+        self.encoding = 546
+        self.charset = 1208
+        self.fmt = ' ' * 8
+        self.flags = 0
+        self.name_value_ccsid = 1208
```

Comment:
does everything between line 134-141 need to be class variables?
most of them looks like they can be constants, and none of them seems to be used outside of `_builder_header`.

Reply:
Nope! :)